### PR TITLE
update `tsh proxy kube` cluster selection ux

### PR DESCRIPTION
**Pull request by GavinFrazar**
This PR allows a user to run `tsh proxy kube` and choose kube cluster(s) by labels, query predicate, name, and/or prefix of name.
Also fixes the `--cluster` flag not being propagated.

`tsh proxy kube` already allowed multiple names to be specified, e.g. `tsh proxy kube name1 name2 name3` - this PR will allow matching those names by unambiguous prefix individually. If any of them are ambiguous, it's an error, which is consistent with other commands.

Part of the implementation for [RFD 129](https://github.com/gravitational/teleport/blob/master/rfd/0129-discovery-name-templating.md).

Related issue:
- https://github.com/gravitational/teleport/issues/22438

Changelog: `tsh proxy kube` now supports `--labels` and `--query` optional arguments.

### CODE REVIEW

File: tool/tsh/common/kube_proxy.go [hunk 1]
```diff
@@ -497,6 +520,39 @@ func issueKubeCert(ctx context.Context, tc *client.TeleportClient, proxy *client
...
+	var out types.KubeClusters
+	for _, clusters := range matchMap {
+		out = append(out, clusters...)
+	}
+	return types.DeduplicateKubeClusters(out)
+}
+
+// matchClustersByNames maps each name to the clusters it matches by exact name
+// or by prefix.
+func matchClustersByNames(cf *CLIConf, tc *client.TeleportClient, clusters types.KubeClusters, names ...string) map[string]types.KubeClusters {
```
Comment (smallinsky) [inline tool/tsh/common/kube_proxy.go:548]: `cf` param is unused: 
```suggestion
func matchClustersByNames(tc *client.TeleportClient, clusters types.KubeClusters, names ...string) map[string]types.KubeClusters {
```

File: tool/tsh/common/kube_proxy.go [hunk 2]
```diff
@@ -497,6 +520,39 @@ func issueKubeCert(ctx context.Context, tc *client.TeleportClient, proxy *client
...
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+// combineMatchedClusters combineMatchedClusters takes a map from name selector
+// to matched clusters and combines all the matched clusters into a deduplicated
+// slice.
+func combineMatchedClusters(cf *CLIConf, tc *client.TeleportClient, matchMap map[string]types.KubeClusters) types.KubeClusters {
```
Comment (smallinsky) [inline tool/tsh/common/kube_proxy.go:538]:
```suggestion
func combineMatchedClusters(matchMap map[string]types.KubeClusters) types.KubeClusters {
```

File: tool/tsh/common/kube_proxy_test.go [hunk 3]
```diff
@@ -86,6 +89,176 @@ func (p *kubeTestPack) testProxyKube(t *testing.T) {
...
+			leafClusters, err := s.leaf.GetAuthServer().GetKubernetesServers(ctx)
+			require.NoError(t, err)
+			return len(rootClusters) == 3 && len(leafClusters) == 1
+		}),
+	)
+	rootClusterName := s.root.Config.Auth.ClusterName.GetClusterName()
+	leafClusterName := s.leaf.Config.Auth.ClusterName.GetClusterName()
+	mustLoginSetEnv(t, s)
+	// Set default kubeconfig to a non-exist file to avoid loading other things.
+	t.Setenv("KUBECONFIG", path.Join(os.Getenv(types.HomeEnvVar), uuid.NewString()))
```
Comment (smallinsky) [inline tool/tsh/common/kube_proxy_test.go:130]:
```suggestion
	t.Setenv(teleport.KubeConfigFile, path.Join(os.Getenv(types.HomeEnvVar), uuid.NewString()))
```

**Review by smallinsky — APPROVED** [review]
LGTM. 
@GavinFrazar Can you postpone  backporting this PR to v13 after we will complete test plan v14 regression ?

File: tool/tsh/common/kube_proxy_test.go [hunk 4]
```diff
@@ -86,6 +89,176 @@ func (p *kubeTestPack) testProxyKube(t *testing.T) {
...
+		{
+			desc: "with labels",
+			validateCmd: func(cmd *exec.Cmd) error {
+				config := kubeConfigFromCmdEnv(t, cmd)
+				checkKubeLocalProxyConfig(t, s, config, rootClusterName, kubeFoo)
+				checkKubeLocalProxyConfig(t, s, config, rootClusterName, kubeFooBar)
+				checkKubeLocalProxyConfig(t, s, config, rootClusterName, kubeBaz)
+				return nil
+			},
+			args: []string{"proxy", "kube", "--labels", "env=root", "--insecure"},
```
Comment (greedy52) [inline tool/tsh/common/kube_proxy_test.go:166]: this use case is awesome

**Review by greedy52 — APPROVED** [review]

File: tool/tsh/common/kube_proxy.go (hunk 2 above)
Comment (GavinFrazar) [reply tool/tsh/common/kube_proxy.go:538]: woops, thanks!

Comment (GavinFrazar) [reply tool/tsh/common/kube_proxy.go:538]: done

File: tool/tsh/common/kube_proxy_test.go (hunk 3 above)
Comment (GavinFrazar) [reply tool/tsh/common/kube_proxy_test.go:130]: teleport.EnvKubeConfig*

changed to use the constant, thank you

**Comment by GavinFrazar** [comment]
@zmb3 can we skip the flaky test detector on this one? I can't speedup `TestProxyKubeComplexSelectors` any further

**Pull request merged**